feat(httpclient): support appending the same header multiple times

sathieu opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana-plugin-sdk-go/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

4. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

-->

**What this PR does / why we need it**:

See https://github.com/grafana/grafana-plugin-sdk-go/issues/829


**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #829

**Special notes for your reviewer**:

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana-plugin-sdk-go?pullRequest=830) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana-plugin-sdk-go?pullRequest=830) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana-plugin-sdk-go?pullRequest=830) it.</sub>

sathieu committed:
feat(httpclient): support appending the same header multiple times

See https://github.com/grafana/grafana-plugin-sdk-go/issues/829

sathieu commented:
@marefr I've fixed the test. back to you 🏸 !

andresmgot commented on backend/http_settings.go:
```diff
@@ -16,7 +16,7 @@ type HTTPSettings struct {
 	BasicAuthUser     string
 	BasicAuthPassword string
-	Headers           map[string]string
+	Headers           map[string][]string
```
I would just go ahead and define this as a `http.Header`

sathieu replied:
Good idea @andresmgot.

I have done that, and renamed from `Headers` to `Header`, as I broke API anyway (and this is now same as `net/http`.

andresmgot reviewed (commented):
I think that in general, it makes sense to use the same type than the upstream HTTP header: https://pkg.go.dev/net/http#Header but I am worried about the breaking change. If we do this, we should prepare at least an adapting PR for grafana/grafana.

Regarding the original issue described in https://github.com/grafana/grafana-plugin-sdk-go/issues/829, I wonder if a possible workaround would be to define the multiple headers as a comma-separated list. @sathieu would that work?

sathieu commented:
Thanks for your review @andresmgot.

> I think that in general, it makes sense to use the same type than the upstream HTTP header: https://pkg.go.dev/net/http#Header but I am worried about the breaking change. If we do this, we should prepare at least an adapting PR for grafana/grafana.

I prepared a PR in grafana at https://github.com/grafana/grafana/pull/79543.

> Regarding the original issue described in #829, I wonder if a possible workaround would be to define the multiple headers as a comma-separated list. @sathieu would that work?

Unfortunately, comma-separated header doesn't work for `prom-label-proxy` (see [code](https://github.com/prometheus-community/prom-label-proxy/blob/0664aa864a4e8aa5858d2772a0576cb3e385bb1f/injectproxy/routes.go#L231-L252)). So this PR is needed.

Back to you!

andresmgot commented on backend/http_settings.go:
```diff
@@ -273,7 +274,7 @@ func parseHTTPSettings(jsonData json.RawMessage, secureJSONData map[string]strin
 		if key, exists := dat[headerNameSuffix]; exists {
 			if value, exists := secureJSONData[headerValueSuffix]; exists {
-				s.Headers[key.(string)] = value
+				s.Header[key.(string)] = append(s.Header[key.(string)], value)
```
I'd change this to:
```go
s.Header.Add(key.(string), value)
```
This has the side effect that the case of the header keys are changed (so you need to adapt the test), but that's fine since keys in HTTP headers are case insensitive.

sathieu replied:
@andresmgot Done. Back to you ⚽ !

andresmgot reviewed (commented):
Thanks, one minor comment here

andresmgot reviewed (approved):
LGTM, any other thought @marefr ?

academo commented:
@andresmgot @marefr is there anything blocking this PR from merging?

marefr commented on backend/http_settings.go:
```diff
@@ -273,7 +274,7 @@ func parseHTTPSettings(jsonData json.RawMessage, secureJSONData map[string]strin
 		if key, exists := dat[headerNameSuffix]; exists {
 			if value, exists := secureJSONData[headerValueSuffix]; exists {
-				s.Headers[key.(string)] = value
+				s.Header.Add(key.(string), value)
```
Don't understand why key needs to be cast to a string - isn't it already a string since `dat` is `map[string]interface{}`?

sathieu replied:
`dat` is `map[string]interface{}`, so `key := dat[headerNameSuffix]` is `interface{}`.

On the other side `secureJSONData` is `map[string]string`, so `value doesn't need cast.

marefr replied:
Oh yeah I see that now :+1:

marefr reviewed (approved):
LGTM, but a question before merging

marefr commented:
@sathieu please sync with main branch to resolve conflicts and see my comment. Thanks

sathieu commented:
@marefr I've rebased and resolved conflicts. Casting is needed.

Back to you 🏓

andresmgot commented:
hi @sathieu, there are some other files that need an update: https://drone.grafana.net/grafana/grafana-plugin-sdk-go/1981/1/3

Also, can you re-open / refresh https://github.com/grafana/grafana/pull/79543? (sorry, it got closed due to inactivity)

marefr commented:
@sathieu back to you. Not sure whether we really should remove `Headers` and add `Header`. I would suggest to keep `Headers`. WDYT? Also, lint problems

```
mage -v lint
go: downloading github.com/magefile/mage v1.15.0
Running target: Lint
exec: golangci-lint "run" "./..."
backend/httpclient/basic_auth_middleware.go:1: : # github.com/grafana/grafana-plugin-sdk-go/backend/httpclient [github.com/grafana/grafana-plugin-sdk-go/backend/httpclient.test]
backend/httpclient/custom_headers_middleware_test.go:67:48: unknown field Headers in struct literal of type Options (typecheck)
package httpclient
backend/httpclient/contextual_middleware_example_test.go:8:2: could not import github.com/grafana/grafana-plugin-sdk-go/backend/httpclient (-: # github.com/grafana/grafana-plugin-sdk-go/backend/httpclient [github.com/grafana/grafana-plugin-sdk-go/backend/httpclient.test]
backend/httpclient/custom_headers_middleware_test.go:67:48: unknown field Headers in struct literal of type Options) (typecheck)
	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
	^
Error: running "golangci-lint run ./..." failed with exit code 1
``

sathieu commented:
@andresmgot :

> hi @sathieu, there are some other files that need an update: https://drone.grafana.net/grafana/grafana-plugin-sdk-go/1981/1/3

Fixed, rebased and pushed.

> Also, can you re-open / refresh [grafana/grafana#79543](https://github.com/grafana/grafana/pull/79543)? (sorry, it got closed due to inactivity)

I can't reopen. I've created a new PR: https://github.com/grafana/grafana/pull/84193

sathieu commented:
@marefr

I think `Header` is more appropriate, like [http.Request](https://pkg.go.dev/net/http#Request), but I can change to `Headers` if you want.

marefr reviewed (approved):
LGTM sure let's use Header
